File: 프로그래머스/lv2/1844. 게임 맵 최단거리/게임 맵 최단거리.py
from collections import deque
def solution(maps):
    answer = 0
    
    dx = [-1, 1, 0, 0]
    dy = [0, 0, -1, 1]
    
    n = len(maps)
    m = len(maps[0])
    isvisited = [[-1 for _ in range(m)] for _ in range(n)]
    q = deque()
    q.append((0, 0))
    isvisited[0][0] = 1
    while q:
        x, y = q.popleft()
        if (x, y) == (n - 1, m - 1):
            return isvisited[x][y]
        for k in range(4):
            nx = x + dx[k]
            ny = y + dy[k]
            if not(0 <= nx < n and 0 <= ny < m): continue
            if maps[nx][ny] == 0: continue
            if isvisited[nx][ny] != -1: continue
            isvisited[nx][ny] = isvisited[x][y] + 1
            q.append((nx, ny))
    return -1

File: 프로그래머스/lv2/1844. 게임 맵 최단거리/게임 맵 최단거리.js
function solution(maps) {
    var answer = 0;
    const row = maps.length;
    const col = maps[0].length;
    let is_visited = Array.from(Array(row), () => Array(col).fill(false));
    is_visited[0][0] = true;
    const dx = [-1, 1, 0, 0];
    const dy = [0, 0, -1, 1];
    const start = [0, 0, 1];
    let q = [start];
    while (q.length > 0) {
        const [x, y, count] = q.shift();
        for(let k = 0; k < 4; k++){
            let nx = x + dx[k];
            let ny = y + dy[k];
            if(nx >= 0 && nx < row && ny >= 0 && ny < col && maps[nx][ny] === 1 && !is_visited[nx][ny]){
                is_visited[nx][ny] = true;
                q.push([nx, ny, count + 1]);
                if(nx === row - 1 && ny === col - 1){
                    console.log(is_visited);
                    return count + 1;
                }
            }
        }
    }
    return -1;
}